File: scripts/parse-rss.js
const fs = require('fs');
const path = require('path');

function extractBetween(str, start, end) {
  const startIdx = str.indexOf(start);
  if (startIdx === -1) return null;
  const contentStart = startIdx + start.length;
  const endIdx = str.indexOf(end, contentStart);
  if (endIdx === -1) return null;
  return str.substring(contentStart, endIdx);
}

function parseItem(itemXml) {
  const title = extractBetween(itemXml, '<title>', '</title>') || '';
  
  // Try different description formats
  let description = extractBetween(itemXml, '<description><![CDATA[', ']]></description>');
  if (!description) description = extractBetween(itemXml, '<description>', '</description>') || '';
  
  let contentEncoded = extractBetween(itemXml, '<content:encoded><![CDATA[', ']]></content:encoded>');
  if (!contentEncoded) contentEncoded = description;
  
  const summary = extractBetween(itemXml, '<itunes:summary><![CDATA[', ']]></itunes:summary>') || '';
  const pubDate = extractBetween(itemXml, '<pubDate>', '</pubDate>') || '';
  const duration = extractBetween(itemXml, '<itunes:duration>', '</itunes:duration>') || '0';
  const guid = extractBetween(itemXml, '<guid isPermaLink="false">', '</guid>') || '';
  
  // Extract enclosure URL
  const enclosureMatch = itemXml.match(/enclosure url="([^"]+)"/);
  const mp3Url = enclosureMatch ? enclosureMatch[1] : null;
  
  // Extract ID from guid (Buzzsprout-XXXXX)
  const id = guid.replace('Buzzsprout-', '');
  
  // Parse title for guest and company
  const cleanTitle = title.replace(/&amp;/g, '&').replace(/&apos;/g, "'");
  const titleParts = cleanTitle.split(' - ');
  const guest = titleParts[0] ? titleParts[0].trim() : 'Brad Roth';
  const company = titleParts[1] ? titleParts[1].trim() : null;
  const topic = titleParts[2] ? titleParts[2].trim() : null;
  
  // Create slug from title
  const slug = cleanTitle
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-|-$/g, '')
    .substring(0, 80);
  
  // Format duration
  const durationSec = parseInt(duration) || 0;
  const mins = Math.floor(durationSec / 60);
  const secs = durationSec % 60;
  const durationFormatted = `${mins}:${secs.toString().padStart(2, '0')}`;
  
  // Strip HTML from description for plain text
  const plainDescription = (contentEncoded || description)
    .replace(/<[^>]+>/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&apos;/g, "'")
    .replace(/&quot;/g, '"')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/\s+/g, ' ')
    .trim();
  
  return {
    id,
    slug,
    title: cleanTitle,
    guest,
    company,
    topic,
    description: plainDescription,
    descriptionHtml: contentEncoded || description,
    summary: summary.replace(/&amp;/g, '&').replace(/&apos;/g, "'"),
    pubDate,
    publishedAt: pubDate ? new Date(pubDate).toISOString() : null,
    mp3Url,
    duration: durationSec,
    durationFormatted
  };
}

function main() {
  console.log('Reading RSS feed from file...');
  const feedPath = path.join(__dirname, '../podcasts/feed.xml');
  const rss = fs.readFileSync(feedPath, 'utf8');
  
  // Split by </item> to get individual items
  const itemRegex = /<item>([\s\S]*?)<\/item>/g;
  const items = [];
  let match;
  while ((match = itemRegex.exec(rss)) !== null) {
    items.push(match[1]);
  }
  
  console.log(`Found ${items.length} episodes`);
  
  const episodes = items.map(item => parseItem(item)).filter(e => e.id);
  
  // Sort by date (newest first)
  episodes.sort((a, b) => new Date(b.publishedAt) - new Date(a.publishedAt));
  
  // Save to file
  const outputPath = path.join(__dirname, '../src/data/episodes.json');
  fs.writeFileSync(outputPath, JSON.stringify(episodes, null, 2));
  
  console.log(`Saved ${episodes.length} episodes to ${outputPath}`);
  
  // Also save MP3 URLs for downloading
  const mp3List = episodes.map(e => ({
    id: e.id,
    slug: e.slug,
    url: e.mp3Url,
    guest: e.guest
  }));
  
  const mp3Path = path.join(__dirname, '../podcasts/mp3-list.json');
  fs.writeFileSync(mp3Path, JSON.stringify(mp3List, null, 2));
  
  console.log(`Saved MP3 list to ${mp3Path}`);
  
  // Print first episode for verification
  console.log('\nFirst episode:');
  console.log(JSON.stringify(episodes[0], null, 2));
}

main();

File: content/podcasts/parse_rss.py
#!/usr/bin/env python3
"""Parse RSS feed and create manifest.json"""
import xml.etree.ElementTree as ET
import json
import re

# Parse the RSS feed
tree = ET.parse('/Users/bradleyroth/clawd/thorfunds-new/content/podcasts/rss_feed.xml')
root = tree.getroot()

# Namespace handling
ns = {
    'itunes': 'http://www.itunes.com/dtds/podcast-1.0.dtd',
    'content': 'http://purl.org/rss/1.0/modules/content/'
}

episodes = []
channel = root.find('channel')

for item in channel.findall('item'):
    # Get itunes:title or regular title
    itunes_title = item.find('itunes:title', ns)
    regular_title = item.find('title')
    title = itunes_title.text if itunes_title is not None and itunes_title.text else (regular_title.text if regular_title is not None else '')
    
    # Get description from itunes:summary
    summary = item.find('itunes:summary', ns)
    description = summary.text[:500] if summary is not None and summary.text else ''
    
    # Get MP3 URL from enclosure
    enclosure = item.find('enclosure')
    mp3_url = enclosure.get('url') if enclosure is not None else ''
    
    # Get pub date
    pub_date = item.find('pubDate')
    date_str = pub_date.text if pub_date is not None else ''
    
    # Get duration
    duration = item.find('itunes:duration', ns)
    duration_secs = int(duration.text) if duration is not None and duration.text and duration.text.isdigit() else 0
    duration_mins = duration_secs // 60
    duration_str = f"{duration_mins} min"
    
    # Get GUID for ID
    guid = item.find('guid')
    guid_text = guid.text if guid is not None else ''
    episode_id = re.search(r'Buzzsprout-(\d+)', guid_text)
    episode_id = episode_id.group(1) if episode_id else ''
    
    # Parse guest and company from title
    if title:
        parts = title.split(' - ')
        guest = parts[0].strip()
        company = parts[1].strip() if len(parts) > 1 else ''
    else:
        guest = ''
        company = ''
    
    if episode_id and mp3_url:
        episodes.append({
            "id": episode_id,
            "title": title,
            "guest": guest,
            "company": company,
            "date": date_str,
            "duration": duration_str,
            "mp3Url": mp3_url,
            "description": description
        })

print(f"Found {len(episodes)} episodes")

# Save manifest
with open('/Users/bradleyroth/clawd/thorfunds-new/content/podcasts/manifest.json', 'w') as f:
    json.dump(episodes, f, indent=2)

print("Manifest saved to manifest.json")
